libraries/apollo-debug-server/src/commonMain/kotlin/com/apollographql/apollo3/debugserver/internal/graphql/GraphQL.kt

```diff
@@ -0,0 +1,219 @@
+package com.apollographql.apollo3.debugserver.internal.graphql
+
+import com.apollographql.apollo3.ApolloClient
+import com.apollographql.apollo3.annotations.ApolloAdapter
+import com.apollographql.apollo3.annotations.ApolloObject
+import com.apollographql.apollo3.annotations.GraphQLName
+import com.apollographql.apollo3.api.Adapter
+import com.apollographql.apollo3.api.CustomScalarAdapters
+import com.apollographql.apollo3.api.ExecutionContext
+import com.apollographql.apollo3.api.json.JsonReader
+import com.apollographql.apollo3.api.json.JsonWriter
+import com.apollographql.apollo3.api.json.writeObject
+import com.apollographql.apollo3.ast.toGQLDocument
+import com.apollographql.apollo3.ast.toSchema
+import com.apollographql.apollo3.cache.normalized.api.CacheKey
+import com.apollographql.apollo3.cache.normalized.api.Record
+import com.apollographql.apollo3.cache.normalized.apolloStore
+import com.apollographql.apollo3.debugserver.internal.graphql.execution.ApolloDebugServerAdapterRegistry
+import com.apollographql.apollo3.debugserver.internal.graphql.execution.ApolloDebugServerResolver
+import com.apollographql.apollo3.execution.ExecutableSchema
+import com.apollographql.apollo3.execution.GraphQLRequest
+import com.apollographql.apollo3.execution.GraphQLRequestError
+import com.apollographql.apollo3.execution.parsePostGraphQLRequest
+import okio.Buffer
+import kotlin.reflect.KClass
+
+internal expect fun KClass<*>.normalizedCacheName(): String
+
+internal expect fun getExecutableSchema(): String
+
+internal class GraphQL(
+    private val apolloClients: Map<ApolloClient, String>,
+) {
+  private val executableSchema: ExecutableSchema by lazy {
+    val schema = getExecutableSchema()
+        .toGQLDocument()
+        .toSchema()
+
+    ExecutableSchema.Builder()
+        .schema(schema)
+        .resolver(ApolloDebugServerResolver())
+        .adapterRegistry(ApolloDebugServerAdapterRegistry)
+        .build()
+  }
+
+  suspend fun executeGraphQL(jsonBody: String): String {
+    val graphQLRequestResult = Buffer().writeUtf8(jsonBody).parsePostGraphQLRequest()
+    if (graphQLRequestResult is GraphQLRequestError) {
+      return graphQLRequestResult.message
+    }
+    graphQLRequestResult as GraphQLRequest
+    val dumps = apolloClients.mapValues { (apolloClient, _) ->
+      apolloClient.apolloStore.dump()
+    }
+    val apolloDebugContext = ApolloDebugContext(apolloClients, dumps)
+    val graphQlResponse = executableSchema.execute(graphQLRequestResult, apolloDebugContext)
```

Comment:
Note to self: would be nice being able to pass arguments to the `Query()` root type instead of relying on executionContext

Reply:
Done in https://github.com/apollographql/apollo-kotlin/pull/5352